src/Storages/MergeTree/CommonANNIndexes.cpp

```diff
@@ -0,0 +1,508 @@
+            function->name == "LpDistance")
+        {
+            out.function = RPNElement::FUNCTION_DISTANCE;
+        }
+        else if (function->name == "tuple")
+        {
+            out.function = RPNElement::FUNCTION_TUPLE;
+        }
+        else if (function->name == "less" ||
+                 function->name == "greater" ||
+                 function->name == "lessOrEquals" ||
+                 function->name == "greaterOrEquals")
+        {
+            out.function = RPNElement::FUNCTION_COMPARISON;
+        }
+        else
+        {
+            return false;
+        }
+
+        return true;
+    }
+    // Match identifier
+    else if (const auto * identifier = node->as<ASTIdentifier>())
+    {
+        out.function = RPNElement::FUNCTION_IDENTIFIER;
+        out.identifier.emplace(identifier->name());
+        out.func_name = "column identifier";
+
+        return true;
+    }
+
+    // Check if we have constants behind the node
+    return tryCastToConstType(node, out);
+}
+
+bool ANNCondition::tryCastToConstType(const ASTPtr & node, RPNElement & out)
+{
+    Field const_value;
+    DataTypePtr const_type;
+
+    if (KeyCondition::getConstant(node, block_with_constants, const_value, const_type))
+    {
+        /// Check for constant types
+        if (const_value.getType() == Field::Types::Float64)
+        {
+            out.function = RPNElement::FUNCTION_FLOAT_LITERAL;
+            out.float_literal.emplace(const_value.get<Float32>());
+            out.func_name = "Float literal";
+            return true;
+        }
+        if (const_value.getType() == Field::Types::UInt64)
+        {
+            out.function = RPNElement::FUNCTION_INT_LITERAL;
+            out.int_literal.emplace(const_value.get<UInt64>());
+            out.func_name = "Int literal";
+            return true;
+        }
+        if (const_value.getType() == Field::Types::Int64)
+        {
+            out.function = RPNElement::FUNCTION_INT_LITERAL;
+            out.int_literal.emplace(const_value.get<Int64>());
+            out.func_name = "Int literal";
+            return true;
+        }
+        if (const_value.getType() == Field::Types::Tuple)
+        {
+            out.function = RPNElement::FUNCTION_LITERAL_TUPLE;
+            out.tuple_literal = const_value.get<Tuple>();
+            out.func_name = "Tuple literal";
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void ANNCondition::traverseOrderByAST(const ASTPtr & node, RPN & rpn)
+{
+    if (const auto * expr_list = node->as<ASTExpressionList>())
+    {
+        if (const auto * order_by_element = expr_list->children.front()->as<ASTOrderByElement>())
+        {
+            traverseAST(order_by_element->children.front(), rpn);
+        }
+    }
+}
+
+bool ANNCondition::matchAllRPNS()
+{
+    ANNExpression expr_prewhere;
+    ANNExpression expr_where;
+    ANNExpression expr_order_by;
+    LimitExpression expr_limit;
```

Comment:
Cosmetic: + newline after expr_limit;

Also below, please put in some newlines after each if-block.